tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb

```diff
@@ -0,0 +1,599 @@
+        "  </td>\n",
+        "  <td>\n",
+        "    <a target=\"_blank\" href=\"https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb\"><img src=\"https://www.tensorflow.org/images/GitHub-Mark-32px.png\" />View source on GitHub</a>\n",
+        "  </td>\n",
+        "  <td>\n",
+        "    <a href=\"https://storage.googleapis.com/tensorflow_docs/model-optimization/tensorflow_model_optimization/g3doc/guide/pruning/pruning_with_sparsity_2_by_4.ipynb\"><img src=\"https://www.tensorflow.org/images/download_logo_32px.png\" />Download notebook</a>\n",
+        "  </td>\n",
+        "</table>"
+      ],
+      "metadata": {
+        "id": "MfBg1C5NB3X0"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Welcome to the guide on the structural pruning M by N.\n",
+        "\n",
+        "Before reading this tutorial it is recommended to get familiar with the concept of pruning and APIs for random pruning:\n",
+        "*  General overview of the pruning technique for the model optimization, see the [overview](https://www.tensorflow.org/model_optimization/guide/pruning).\n",
+        "*  Usage of API's on a single end-to-end example, see the [pruning example](https://www.tensorflow.org/model_optimization/guide/pruning/pruning_with_keras).\n",
+        "\n",
+        "In this tutorial, you will:\n",
+        "* Define and train a model on the mnist dataset with structural sparsity 2 by 4\n",
+        "* Convert the pruned model to tflite format\n",
+        "* Visualize structure of the pruned weights\n"
+      ],
+      "metadata": {
+        "id": "FbORZA_bQx1G"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Structural pruning M by N"
+      ],
+      "metadata": {}
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Structural pruning zeroes out model weights at the beginning of the training\n",
+        "process according to the following pattern: M weights are set to zero in the\n",
+        "block of N weights. It is important to notice that this pattern affects only the last dimension of the weight tensor for the model that is converted by TensorFlow Lite. For example, `Conv2D` layer weights in TensorFlow Lite have the structure [channel_out, height, width, channel_in] and `Dense` layer weights have the structure [channel_out, channel_in]. The sparsity pattern is applied to the weights in the last dimension: channel_in.\n",
+        "Special hardware can benefit from this type of sparsity in the model and inference time can have a significant speedup. Because this pattern lock in sparsity is more restrictive, the accuracy achieved after fine-tuning is worse than with the magnitude-based pruning.\n",
+        "It is important to indicate that the pattern is valid only for the model that is converted to tflite.\n",
```

Comment:
HI @wwwind, 

While internal review, I found I don't understand this sentence. Could you explain it with reply?

Reply:
Hi @rino20 

If we check the model that is not saved to tflite format, then we don't see this pattern in the last dimension. 
Because, for Conv2D, for example: internally in TF data format we have (h, w, channel_in(kernel_size), channel_out(filters)). when we save into the tflite data format, we have: (channel_out, h, w, channel_in). The sparsity pattern is in the last dimension.